### ISSUE TITLE
NC method error

### ISSUE THREAD

**Issue by avdvg** [Unispac/Circumventing-Backdoor-Defenses#1]
Dear developer:

Hello~ In the process of running the NC poisoning detection code, I found that the entropy value of the label 0 is very low. In the process of reviewing the code, I found that in your code of Circumventing-Backdoor-Defenses/other_defenses/neural_cleanse.py, there seems to be an error in the input for calculating the cross-entropy. I think the input for calculating the cross-entropy should be a one-hot label ( The corresponding position is 1), instead of directly multiplying the number of labels with the tensor of all 1.

```
def loss_fn(self, _input, _label, Y, mask, mark, label):
        X = (_input + mask * (mark - _input)).clamp(0., 1.)
        Y = label * torch.ones_like(_label, dtype=torch.long)
        _output = self.model(self.normalizer(X))
        return self.criterion(_output, Y)
```

Then, I modified the code as follows and it worked
```
    def loss_fn(self, _input, _label, Y, mask, mark, label):
        X = (_input + mask * (mark - _input)).clamp(0., 1.) 
        Y = torch.zeros_like(_label) 
        Y[:,label] = 1
        _output = self.model(self.normalizer(X))
        return self.criterion(_output, Y)
```

**Comment by vtu81** [comment]
Hi avdvg,

Thanks for your participation into our development! I believe our implementation here is correct. The `_label` here is a one-dimensional tensor of shape `(batch_size,)`, where each element is an integer indicating the corresponding input's original label.  (Notice the input parameter `Target` of `nn.CrossEntropyLoss()` should usually be a tensor of shape `(N,)`, but not an one-hot label tensor, see [here](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html).) NeuralCleanse aims at reverse engineering a trigger for each class (`label`) by optimizing on the input space and re-label all inputs to `label`, regardless of their own ground-truth class. Therefore, we multiply the tensor `torch.ones_like(_label)` with the currently suspected `label`, to obtain a tensor like `[label, label, label, ..., label]` as the optimization target.

I also tried running your code, which led to a runtime error while calculating cross-entropy. I presume you are using a deprecated version of PyTorch? Thanks for your issue anyway!!
